File: solutions/Sliding Window/121_best-time-to-buy-and-sell-stock.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        minn = float("inf")
        res = 0

        for i in prices:
            if i < minn:
                minn = i
            else:
                res = max(res, i - minn)

        return res
    
# Time Complexity: O(n)
# Space Complexity: O(1)

File: C++/Sliding Window/121_best-time-to-buy-and-sell-stock.cpp
#include <vector>
#include <climits>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int minValue = INT_MAX;
        int res = 0;

        for (int price: prices) {
            if (price < minValue) minValue = price;
            else res = max(res, price - minValue);
        }

        return res;
    }
};